Add more taxonomy options to the post navigation link

carolinan opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Updated Nov 6 2023.

This PR adds support for two parameters of [get_next_post_link](https://developer.wordpress.org/reference/functions/get_next_post_link/) / [get_previous_post_link](https://developer.wordpress.org/reference/functions/get_previous_post_link/) that were not supported by the block.

$in_same_term bool Optional
Whether link should be in a same taxonomy term.
Default: false

$taxonomy string Optional
Taxonomy, if $in_same_term is true. Default 'category'.
Default: 'category'

- The feature for excluding terms was initially included, but has been removed based on feedback that it was too confusing.

Closes https://github.com/WordPress/gutenberg/issues/47090

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
The features are available in classic themes but were missing from the block. The feature allows site creators and developers to limit what should be included as a next and previous link.

This allows site creators to for example add a 'news' category to some posts, and only link to the other posts in this category.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
Adds a select list with taxonomies under the Advanced panel. Selecting a taxonomy in this list will limit the next and previous links.

To get a list of all supported taxonomies, including custom taxonomies, we need to know the post type.
I was not able to find any other way to get the post type besides using the core editor store.
If you know of a better way please comment on the PR.

## Testing Instructions
Create three identical posts:
Add the posts to the "birds" category, or any new category that you want to add for the test.
Add the next and previous post navigation link blocks. Open the advanced panel and select category as the taxonomy.

View the first post on the front.
You should be able to navigate to the second post in the birds category, and from there, to the third post in the birds category.
When viewing the third post, the "next" link should not show.

Register a custom post type and custom taxonomy. The fastest way to do this is to enable a plugin with custom post types, I used https://wordpress.org/plugins/yoast-test-helper/.
Repeat the testing instructions above but use the custom post type and custom taxonomy (movie and genre for example).
Only the taxonomies supported by the post type should show in the taxonomy option.

Next, with a block theme active, please test the block in the Site Editor:
Go to Appearance > Editor > Templates, and edit the single post template.
Select or add the post navigation link blocks, and enable the filter under the Advanced panel.
Save the template changes.
Confirm that the placeholder in the Site Editor is the same, and that the correct posts are linked on the front of the website.

## Screenshots or screencast <!-- if applicable -->
<img width="224" alt="Filter option under the Advanced panel" src="https://github.com/WordPress/gutenberg/assets/7422055/0b8133d9-a136-4e2f-8b63-15174e5cd800">

carolinan committed:
Add 3 options to the post navigation link

carolinan committed:
Update core__post-navigation-link.json

carolinan committed:
Support custom post types and custom taxonomies

carolinan committed:
fix CS issues

carolinan commented:
This is still in draft because I am testing an issue where the "excluded terms" does not seem to fully reset if the user empties the input field.

carolinan committed:
Update index.php

carolinan committed:
Update index.php

mikachan commented on packages/block-library/src/post-navigation-link/block.json:
```diff
@@ -28,6 +28,16 @@
 			"default": "none"
+		},
+		"excludedTerms": {
+			"type": "string"
```
```suggestion
			"type": [ "string", "array" ]
```

It looks like `excludedTerms` can also be an array, should we include this in the accepted types here?

draganescu replied:
Should this only apply as to what the attribute of the block allows, now to what the `get_next_post` WP function allows? In that case if the code always works with strings (coming from the inspector input) no reason to support array I think.

mikachan replied:
Ahh ok, that makes sense. I guess multiple values would be comma-separated in the inspector input string then.

mikachan commented on packages/block-library/src/post-navigation-link/edit.js:
```diff
@@ -47,6 +61,42 @@ export default function PostNavigationLinkEdit( {
+			context: 'view',
+		} );
+		return filteredTaxonomies;
+	}, [] );
```
```suggestion
	}, [ postType ] );
```

I think this hook has a dependency on `postType`, so it should be included in the dependency array.

mikachan reviewed (commented):
I've tested this out and it's working well for me 🎉

- You should be able to navigate to the second post in the birds category, and from there, to the third post in the birds category. ✅
- When viewing the third post, the "next" link should not show. ✅
- The next link should link to post number four that you just created, post 2 and 3 should be excluded. ✅
- Only the taxonomies supported by the post type should show in the taxonomy option. ✅

I've left some minor inline comments.

richtabor commented:
Perhaps the toggle label should be “Filter by taxonomy”

carolinan commented:
> Perhaps the toggle label should be “Filter by taxonomy”

Yes! Much better.

carolinan committed:
Update edit.js

mikachan reviewed (approved):
Looks good to me!

scruffian commented:
I rebased this

scruffian commented:
> Perhaps the toggle label should be “Filter by taxonomy”

I disagree. This made me think that the links would be filtered by whatever string was in the input field!

carolinan committed:
Merge branch 'trunk' into update/post-navigation-link

carolinan committed:
Merge branch 'trunk' into update/post-navigation-link

carolinan committed:
try to fix spacing in core-blocks.md

carolinan committed:
Merge branch 'trunk' into update/post-navigation-link

carolinan commented:
What if I remove the toggle option and only show the taxonomy select list? The default option would be "All".

carolinan committed:
Merge branch 'trunk' into update/post-navigation-link

carolinan committed:
Remove the taxonomy filter toggle and update the PHP conditions.

carolinan commented:
I have removed the toggle and tweaked the logic a bit.

Before: You would set the `inSameTerm` block attribute to true with the toggle option and then select a value for the `taxonomy` attribute.

After: Now, selecting a taxonomy updates both the `inSameTerm` and `taxonomy` attributes.
The default option in the select list is not "All" but "Unfiltered".

I added a new help text: "Only link to posts that have the same taxonomy terms as the current post. For example the same categories."

https://github.com/WordPress/gutenberg/assets/7422055/373aa2c2-3f5b-4cf6-a6b4-619424a38d65

carolinan committed:
Update core__post-navigation-link.json

scruffian commented:
Thanks for updating this. I reviewed it again, and there are still a few things I find confusing.

When I select "Filter by taxonomy" as "Unfiltered", I still see "Add terms to exclude". I assume that this input doesn't do anything in this case? If so we should hide it.

If I select "tags", I assume that link will point only to posts that have all the same tags as the current post. However if I enter a tag to exclude which is part of the current post, I find that the link never displays anything even if there are other posts that share the other tags. Is that expected?

carolinan commented:
The excluded terms can be in any taxonomy and it is unrelated to the taxonomy in the filter,  see https://developer.wordpress.org/reference/functions/get_adjacent_post/

I need to test and try to the tag issues. It might be something in the logic in the PHP file.

carolinan committed:
Merge branch 'trunk' into update/post-navigation-link

carolinan committed:
Update help text about excluding terms

carolinan commented:
> If I select "tags", I assume that link will point only to posts that have all the same tags as the current post. However if I enter a tag to exclude which is part of the current post, I find that the link never displays anything even if there are other posts that share the other tags. Is that expected?

I was able to reproduce this. I find the expectation to be unclear in the documentation, but I believe this is the query used:
https://core.trac.wordpress.org/browser/tags/6.3/src/wp-includes/link-template.php#L1887

```

if ( ! empty( $excluded_terms ) ) {
$where .= " AND p.ID NOT IN ( SELECT tr.object_id FROM $wpdb->term_relationships tr LEFT JOIN $wpdb->term_taxonomy tt ON (tr.term_taxonomy_id = tt.term_taxonomy_id) WHERE tt.term_id IN (" . implode( ',', array_map( 'intval', $excluded_terms ) ) . ') )';
}

```

I did not find a condition that says that if there are excluded terms but there are other matches, include the (post).

carolinan commented:
What we could do is move one or both settings to the Advanced panel.
Also, I would prefer to not have the same arguments as with the excerpt length setting, so I'd like the option to turn this off.

carolinan committed:
Update help text again.

carolinan commented:
I was thinking something like this:

block.json:

```
	"attributes": {
		"textAlign": {
			"type": "string"
		},
		"type": {
			"type": "string",
			"default": "next"
		},
		"label": {
			"type": "string"
		},
		"showTitle": {
			"type": "boolean",
			"default": false
		},
		"linkLabel": {
			"type": "boolean",
			"default": false
		},
		"arrow": {
			"type": "string",
			"default": "none"
		},
		"allowFilters": {.  <----- This is the new attribute
			"type": "boolean",
			"default": true
		},
		"excludedTerms": {
			"type": "string",
			"default": ""
		},
		"inSameTerm": {
			"type": "boolean",
			"default": false
		},
		"taxonomy": {
			"type": "string",
			"default": ""
		}
	},
```

In `Edit.js` there would be a conditional that checks if `allowFilters` is true, and if not, the controls are not visible.

And then to turn the filter off and set defaults:

```
<?php
function filter_metadata_registration( $metadata ) {
	if ( 'core/post-navigation-link' === $metadata['name'] ) {
		$metadata['attributes']['allowFilters'] = array(
			'type'    => 'boolean',
			'default' => false,
		);
		$metadata['attributes']['excludedTerms'] = array(
			'type'    => 'string',
			'default' => 'uncategorized',
		);
		$metadata['attributes']['inSameTerm'] = array(
			'type'    => 'boolean',
			'default' => 'true',
		);
		$metadata['attributes']['taxonomy'] = array(
			'type'    => 'string',
			'default' => 'category',
		);
	}
	return $metadata;
};
add_filter( 'block_type_metadata', 'filter_metadata_registration' );
```

I haven't gotten it to work yet; If the block already has a term added to the `excludedTerms`, adding another term by using the filter only seems to combine the terms, not reset them.
I mean that turning the UI off with the filter does not reset previous user settings. Should it?

carolinan committed:
Merge branch 'trunk' into update/post-navigation-link

carolinan committed:
Merge branch 'trunk' into update/post-navigation-link

carolinan committed:
Merge branch 'trunk' into update/post-navigation-link

carolinan commented:
@WordPress/block-themers Hi people, I would appreciate your thoughts about this feature and the suggestion from August 11 (above) about allowing it to be disabled.

carolinan committed:
Merge branch 'trunk' into update/post-navigation-link

carolinan committed:
Merge branch 'trunk' into update/post-navigation-link

draganescu commented on packages/block-library/src/post-navigation-link/edit.js:
```diff
@@ -47,6 +60,44 @@ export default function PostNavigationLinkEdit( {
+
+	// We need to know the post type in order to get all supported taxonomies.
+	const postType = useSelect(
+		// FIXME: @wordpress/block-library should not depend on @wordpress/editor.
```
I am not sure this fixme is needed. Also why the eslint exception? I think "post navigation" will require a post context. Some blocks in the block library _are_ inextricably connected to WordPress and it's system.

carolinan replied:
It is consistent with how some other blocks handle this store, but it feels like a temporary solution and it is part of a bigger discussion. I believe this is the main issue: https://github.com/WordPress/gutenberg/issues/32842
@gziolo

gziolo replied:
This won't work as expected in the site editor by default. It would only work correctly if a plugin somehow enqueues `wp-editor`, but it isn't true for Gutenberg and WordPress core.

Have you tried sourcing the same information from the block context by defining: `"usesContext": [ "postType" ],` in `block.json`? Example:

https://github.com/WordPress/gutenberg/blob/bc704e140ce6844ce2dc1666bb3697801970bdc7/packages/block-library/src/post-author-biography/block.json#L14

If the `postType` isn't available in the post editor, it would be a good idea to expose it there through block context and replace all occurrences that currently fetch the same information from the hardcoded `core/editor` store.

carolinan replied:
Thank you, I will try that and reply here.

gziolo replied:
Let me know if there are any issues with the idea. We can always land as is and work on the general solution in the follow-up.

carolinan replied:
_Context_ seems to be working well in both the Site Editor and post editor. But I am testing manually, not sure which edge cases I need to test and how. 🤔

gziolo replied:
https://github.com/WordPress/gutenberg/pull/48912/commits/54a58bb9a10c287ee3cf86c6004705f25d3d4ce9 that looks great :+1:

draganescu reviewed (approved):
I think this is a great addition to this block!

carolinan committed:
Remove the exclude terms feature, move filter to advanced panel

carolinan committed:
Try to fix CS issues

carolinan committed:
Update docs

carolinan committed:
Try to fix CS issues

carolinan committed:
Use context to get the postType

carolinan commented on packages/block-library/src/post-navigation-link/edit.js:
```diff
@@ -47,6 +59,40 @@ export default function PostNavigationLinkEdit( {
+			.filter(
+				( tax ) =>
+					tax.slug !== 'nav_menu' &&
+					tax.slug !== 'wp_pattern_category'
```
The pattern category shows in the option in Site Editor unless it is removed here.

If the "New admin views" Gutenberg experiment is enabled, data views also show in the option. I did not hide it since it is still experimental.

t-hamano replied:
Is this conditional statement really necessary? I hesitate to hardcode specific taxonomy slugs into blocks. For example, in a query loop block, no such condition exists.

https://github.com/WordPress/gutenberg/blob/55b4e9b7f3861b7d782dec4a4b839ace0e30f300/packages/block-library/src/query/utils.js#L130-L144

carolinan replied:
I simply deleted the filter and nothing changed in the interface. I am not sure what changed since November but neither the nav menu, pattern categories or the admin views are showing up now. Custom taxonomies, for standard and for custom post types are showing correctly as well.

t-hamano replied:
I don't know about the navigation menu and admin view, but it seems that `public` has been changed to `false` in the pattern category.

- https://core.trac.wordpress.org/ticket/59569
- #55748

carolinan committed:
Merge branch 'trunk' into update/post-navigation-link

carolinan committed:
Merge branch 'trunk' into update/post-navigation-link

carolinan committed:
Merge branch 'trunk' into update/post-navigation-link

t-hamano commented on packages/block-library/src/post-navigation-link/edit.js:
```diff
@@ -114,6 +160,22 @@ export default function PostNavigationLinkEdit( {
+					value={ taxonomy }
+					options={ getTaxonomyOptions() }
+					onChange={ ( value ) =>
+						setAttributes( {
+							taxonomy: value,
+							inSameTerm: value === '' ? false : true,
+						} )
```
The value of `isSameTerm` depends only on whether the value of `taxonomy` is empty. So, can this attribute be removed and not determined during dynamic rendering?

carolinan commented:
@draganescu can you revert the commit? I'd rather not deal with another deprecation.

carolinan commented:
It is disheartening to work on something for close to a year and not receive code reviews until commit, it has become quiet difficult to find willing reviewers the past 6 months, -and it is not a problem with one pull request but most of them. All I can do is ping people.

carolinan commented:
I want to contribute to the project on issues that matter and are on the roadmap, not work on things that are not worthy peoples time to review.

t-hamano commented:
> can you revert the commit? I'd rather not deal with another deprecation.

If I understand correctly, this block is a dynamic block, so if we don't need to migrate existing attributes, I think we don't need the deprecation, and we don't need to revert. If there is something that can be improved, I think it would be a good idea to commit to it as a follow-up.

carolinan commented:
Then I hope that with both your help I can fix it before 17.6.
